### PR TITLE
Start a time entry from suggestions

### LINKED ISSUE #691: Start timer from suggestions

### PR DESCRIPTION
**Pull request by arturs105**
Closes #691 

Do a functional test before approving

Do you think we should make the whole cell clickable instead of just the button?

### CODE REVIEW

File: Toggl.Foundation.MvvmCross/ViewModels/SuggestionsViewModel.cs [hunk 1]
```diff
@@ -62,5 +75,25 @@ private void addSuggestions(Suggestion suggestions)
...
+        {
+            await dataSource.TimeEntries.Stop(timeService.CurrentDateTime)
+                .Catch((NoRunningTimeEntryException e) => Observable.Return(default(IDatabaseTimeEntry)));
+
+            await dataSource.User
+                .Current()
+                .Select(user => new StartTimeEntryDTO
+                {
+                    UserId = user.Id,
+                    WorkspaceId = user.DefaultWorkspaceId,
```
Comment (simonrozsival) [inline Toggl.Foundation.MvvmCross/ViewModels/SuggestionsViewModel.cs]: This seems strange. I can see that there is no `WorkspaceId` in the `Suggestion` class, but shouldn't there be one? Some suggestions probably won't have one (e.g., suggestion based on calendar), but some might have one (e.g., frequently used time entry).

Comment (amulware) [reply Toggl.Foundation.MvvmCross/ViewModels/SuggestionsViewModel.cs]: Yes, suggestions should always include a valid workspace id, which should be decided by the provider.

File: Toggl.Foundation.MvvmCross/ViewModels/SuggestionsViewModel.cs [hunk 2]
```diff
@@ -62,5 +75,25 @@ private void addSuggestions(Suggestion suggestions)
             Suggestions.Add(suggestions);
             RaisePropertyChanged(nameof(IsEmpty));
         }
+
+        private async Task startTimeEntry(Suggestion suggestion)
+        {
+            await dataSource.TimeEntries.Stop(timeService.CurrentDateTime)
```
Comment (simonrozsival) [inline Toggl.Foundation.MvvmCross/ViewModels/SuggestionsViewModel.cs]: 🔥 this - it is unnecessary, this is handled by the repository.

File: Toggl.Foundation.Tests/MvvmCross/ViewModels/SuggestionsViewModelTests.cs [hunk 3]
```diff
@@ -122,5 +129,140 @@ public sealed class TheSuggestionsProperty : SuggestionsViewModelTest
             private Recorded<Notification<Suggestion>> createRecorded(int ticks, Suggestion suggestion)
                 => new Recorded<Notification<Suggestion>>(ticks, Notification.CreateOnNext(suggestion));
         }
+
+        public sealed class TheStartTimeEntryCommand : SuggestionsViewModelTest
+        {
+            [Fact]
+            public async Task CallsStopBeforeStartingANewTimeEntry()
```
Comment (simonrozsival) [inline Toggl.Foundation.Tests/MvvmCross/ViewModels/SuggestionsViewModelTests.cs]: 🔥 this test

File: Toggl.Foundation.Tests/MvvmCross/ViewModels/SuggestionsViewModelTests.cs [hunk 4]
```diff
@@ -122,5 +129,140 @@ public sealed class TheSuggestionsProperty : SuggestionsViewModelTest
...
+
+                Received.InOrder(async () =>
+                {
+                    await DataSource.TimeEntries.Stop(Arg.Any<DateTimeOffset>());
+                    await DataSource.TimeEntries.Start(Arg.Any<StartTimeEntryDTO>());
+                });
+            }
+
+            [Fact]
+            public async Task StartsATimeEntryEvenIfTheStopMethodThrowsBecauseThereWasNoRunningTimeEntry()
```
Comment (simonrozsival) [inline Toggl.Foundation.Tests/MvvmCross/ViewModels/SuggestionsViewModelTests.cs]: 🔥 this test

File: Toggl.Foundation.Tests/MvvmCross/ViewModels/SuggestionsViewModelTests.cs [hunk 5]
```diff
@@ -122,5 +129,140 @@ public sealed class TheSuggestionsProperty : SuggestionsViewModelTest
...
+            {
+                var suggestion = createSuggestion();
+
+                await ViewModel.StartTimeEntryCommand.ExecuteAsync(suggestion);
+
+                await DataSource.SyncManager.Received().PushSync();
+            }
+
+            [Fact]
+            public async void InitiatesPushSyncWhenThereIsNoRunningTimeEntry()
```
Comment (simonrozsival) [inline Toggl.Foundation.Tests/MvvmCross/ViewModels/SuggestionsViewModelTests.cs]: 🔥 this test

File: Toggl.Foundation.Tests/MvvmCross/ViewModels/SuggestionsViewModelTests.cs [hunk 6]
```diff
@@ -122,5 +129,140 @@ public sealed class TheSuggestionsProperty : SuggestionsViewModelTest
...
+                DataSource.TimeEntries.Received().Start(Arg.Is<StartTimeEntryDTO>(dto =>
+                    dto.Description == description.Get &&
+                    dto.TaskId == taskId &&
+                    dto.ProjectId == projectId &&
+                    dto.WorkspaceId == workspaceId
+                )).Wait();
+            }
+
+            [Fact]
+            public async void InitiatesPushSyncWhenThereIsARunningTimeEntry()
```
Comment (simonrozsival) [inline Toggl.Foundation.Tests/MvvmCross/ViewModels/SuggestionsViewModelTests.cs]: Rename this test. This view model has no idea how stopping a time entry works and it doesn't need to test it.

File: Toggl.Foundation.Tests/MvvmCross/ViewModels/SuggestionsViewModelTests.cs [hunk 7]
```diff
@@ -122,5 +129,140 @@ public sealed class TheSuggestionsProperty : SuggestionsViewModelTest
...
+                DataSource.TimeEntries.Stop(Arg.Any<DateTimeOffset>())
+                    .Returns(Observable.Throw<IDatabaseTimeEntry>(new NoRunningTimeEntryException()));
+
+                await ViewModel.StartTimeEntryCommand.ExecuteAsync(suggestion);
+
+                await DataSource.SyncManager.Received().PushSync();
+            }
+
+            [Fact]
+            public async void DoesNotInitatePushSyncWhenStoppingFailsForOtherReasonThanThatThereIsNoRunningTimeEntry()
```
Comment (simonrozsival) [inline Toggl.Foundation.Tests/MvvmCross/ViewModels/SuggestionsViewModelTests.cs]: 🔥 this test

**Pull request merged**